src/python/pants/backend/typescript/dependency_inference/rules.py

```diff
@@ -0,0 +1,141 @@
+from __future__ import annotations
+
+import itertools
+import os.path
+from dataclasses import dataclass
+from typing import Iterable
+
+from pants.backend.javascript import package_json
+from pants.backend.javascript.dependency_inference.rules import (
+    InferNodePackageDependenciesRequest,
+    NodePackageCandidateMap,
+    RequestNodePackagesCandidateMap,
+)
+from pants.backend.javascript.package_json import (
+    OwningNodePackage,
+    OwningNodePackageRequest,
+    PackageJsonSourceField,
+)
+from pants.backend.javascript.subsystems.nodejs_infer import NodeJSInfer
+from pants.backend.typescript.target_types import (
+    TS_FILE_EXTENSIONS,
+    TypeScriptDependenciesField,
+    TypeScriptSourceField,
+)
+from pants.build_graph.address import Address
+from pants.engine.addresses import Addresses
+from pants.engine.internals.graph import Owners, OwnersRequest
+from pants.engine.internals.native_dep_inference import NativeParsedJavascriptDependencies
+from pants.engine.internals.native_engine import InferenceMetadata, NativeDependenciesRequest
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.rules import Rule, collect_rules, rule
+from pants.engine.target import (
+    FieldSet,
+    HydratedSources,
+    HydrateSourcesRequest,
+    InferDependenciesRequest,
+    InferredDependencies,
+    Targets,
+)
+from pants.engine.unions import UnionRule
+from pants.util.ordered_set import FrozenOrderedSet
+
+
+@dataclass(frozen=True)
+class TypeScriptSourceInferenceFieldSet(FieldSet):
+    required_fields = (TypeScriptSourceField, TypeScriptDependenciesField)
+
+    source: TypeScriptSourceField
+    dependencies: TypeScriptDependenciesField
+
+
+@dataclass(frozen=True)
+class TypeScriptFileImportPath:
+    """Path to a file that is imported in TypeScript code."""
+
+    path: str
+
+
+class InferTypeScriptDependenciesRequest(InferDependenciesRequest):
+    infer_from = TypeScriptSourceInferenceFieldSet
+
+
+async def _prepare_inference_metadata(address: Address) -> InferenceMetadata:
+    owning_pkg = await Get(OwningNodePackage, OwningNodePackageRequest(address))
+    if not owning_pkg.target:
+        return InferenceMetadata.javascript(address.spec_path, {})
+    return await Get(
+        InferenceMetadata, PackageJsonSourceField, owning_pkg.target[PackageJsonSourceField]
+    )
+
+
+@rule
+async def get_file_imports_targets(import_path: TypeScriptFileImportPath) -> Targets:
+    """Get address to build targets representing a file import discovered with dependency inference.
+
+    For now, we only support .ts files; in the future, may need to iterate through all possible file
+    extensions until find one matching, if any.
+    """
+    package, filename = os.path.dirname(import_path.path), os.path.basename(import_path.path)
+    address = Address(package, relative_file_path=f"{filename}{TS_FILE_EXTENSIONS[0]}")
+    _ = await Get(Targets, Addresses([address]))
```

Comment:
What's the point of this line? The `_` isn't referenced later on.

Reply:
If the import attempts to refer to a file that doesn't exist, a `ResolveError` will be raised on attempt to get a target for a given address.

With 

```diff
diff --git a/testprojects/src/ts/frontend/config/app.ts b/testprojects/src/ts/frontend/config/app.ts
index fc71d23618..df31a7ee98 100644
--- a/testprojects/src/ts/frontend/config/app.ts
+++ b/testprojects/src/ts/frontend/config/app.ts
@@ -11,3 +11,5 @@ import * as Sentry from '@sentry/react';
 // local file import
 import { dispatcher } from './dispatcher';
 import { receiver } from './receiver';
+
+import { bla } from './bla';
```

run

```
$ pants dependencies testprojects/src/ts/frontend/config/app.ts
pants.build_graph.address.ResolveError: The address `testprojects/src/ts/frontend/config/bla.ts` from <infallible> was not generated by the target `testprojects/src/ts/frontend/config:config`. Did you mean one of these addresses?

  * testprojects/src/ts/frontend/config:config
  * testprojects/src/ts/frontend/config/app.ts
  * testprojects/src/ts/frontend/config/dispatcher.ts
  * testprojects/src/ts/frontend/config/receiver.ts
```

Let me add a comment to explain the call.

Reply:
Without this extra call, same diff (with `./bla` import`):

```
$ pants dependencies testprojects/src/ts/frontend/config/app.ts
testprojects/src/ts/frontend/config/app.ts
testprojects/src/ts/frontend/config/dispatcher.ts
testprojects/src/ts/frontend/config/receiver.ts
```

Reply:
Or is there a better way to fail fast telling user about the invalid import?

Reply:
See above (https://github.com/pantsbuild/pants/pull/20293/files#r1428401017); I'm definitely forming the view that precomputing the set of importable targets is going to make for cleaner/more understandable code in the long run.

Reply:
I think this can be either 1. left as is or 2. removed and then migrated to what will be part of https://github.com/pantsbuild/pants/pull/21176. That PR will introduce the same option as python inference (fail, ignore and warn on imports pants couldn't parse).

I'll leave that up to you @AlexTereshenkov . My personal preference is for pants not to hard fail when inference fails because pants isnt a linter, when it fails it hard stops processes it orchestrates.